janus_server/src/datastore.rs

```diff
@@ -461,6 +531,27 @@ fn single_row(rows: Vec<Row>) -> Result<Row, Error> {
     }
 }
 
+/// Extensions for [`tokio_postgres::row::Row`]
+trait RowExt {
+    /// Get a PostgreSQL `BIGINT` from the row, which is represented in Rust as
+    /// i64 ([1]), then attempt to convert it to u64.
+    ///
+    /// [1]: https://docs.rs/postgres-types/latest/postgres_types/trait.FromSql.html
+    fn get_bigint_as_u64<I>(&self, idx: I) -> Result<u64, Error>
+    where
+        I: RowIndex + Display;
+}
+
+impl RowExt for Row {
+    fn get_bigint_as_u64<I>(&self, idx: I) -> Result<u64, Error>
+    where
+        I: RowIndex + Display,
+    {
+        let bigint: i64 = self.try_get(idx)?;
+        Ok(u64::try_from(bigint)?)
```

Comment:
I think we really do want to do `nonce_rand as u64` rather than `u64::try_from(nonce_rand)`, when converting `nonce_rand` specifically. (this may not hold for other `u64` values that are stored as `BIGINT` in the DB.)

`u64::try_from(some_i64)` will fail if `some_i64` has a numeric value that is outside the range of what can be represented as `u64`. OTOH, `some_i64 as u64` simply reinterprets the bytes (and so may not give the same numeric value).

Since this part of the nonce is random, about half of the possible `u64` values we might generate will not be representable as an `i64`. Prior to this change, we'd just convert back & forth by reinterpreting the bytes, since we ultimately don't care about the specific numeric value, only that we get a consistent result & can check equality. With this change, we'll error out whenever we happen to generate a random part of the nonce that is outside the numeric range of `i64`.

All that said: this is perhaps too confusing/error-prone. (If nothing else, for a given `nonce_rand` value we'd need to refer to it as one numeric value in Rust, and a different numeric value in Postgres.) Perhaps we should instead represent the random part of the nonce as e.g. `[u8; 8]` in Rust & as a `BYTEA` in the DB? This typing is also semantically closer to what `nonce_rand` actually is, i.e. an opaque, uninterpreted random value.

Rust playground: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d1465a9869f0b23c6fe62b267c0b35ad

Reply:
I like that idea, because TBH all these signed-to-unsigned conversions make me nervous due to past trauma induced by [`CFNumberGetValue`](https://developer.apple.com/documentation/corefoundation/1543114-cfnumbergetvalue?language=objc).